### ISSUE TITLE
systemd creates a new PrivateTmp directory for rngd.service on every daemon-reload

### ISSUE THREAD

**Issue by hexchain** [systemd/systemd#17313]
**systemd version the issue has been seen with**
 > 246

**Used distribution**
 > Arch

**Linux kernel version used** (`uname -a`)
 > Linux host 5.8.10-zen1-1-zen #1 ZEN SMP PREEMPT Thu, 17 Sep 2020 18:01:58 +0000 x86_64 GNU/Linux

**CPU architecture issue was seen on**
 > x86-64

**Expected behaviour you didn't see**
 > systemd should not create new private temporary directories on every reload

**Unexpected behaviour you saw**
 > systemd creates a new private temporary directory for rngd.service on every reload

**Steps to reproduce the problem**
Have `rngd.service` enabled and started, and notice that every `daemon-reload` results in a new directory being created.
Other services utilizing `PrivateTmp` does not seem affected.
```
/tmp % ls -lhtr | tail -n1
drwx------  3 root     root       60 Oct 11 16:33 systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-YPwYuj/
/tmp % date; sudo systemctl daemon-reload
Sun Oct 11 04:34:33 PM CEST 2020
/tmp % ls -lhtr | tail -n2
drwx------  3 root     root       60 Oct 11 16:33 systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-YPwYuj/
drwx------  3 root     root       60 Oct 11 16:34 systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-t7Xich/
/tmp % systemctl cat rngd.service
# /usr/lib/systemd/system/rngd.service
[Unit]
Description=Hardware RNG Entropy Gatherer Daemon
DefaultDependencies=no
Before=sysinit.target shutdown.target

[Service]
EnvironmentFile=/etc/conf.d/rngd
ExecStart=/usr/bin/rngd -f $RNGD_OPTS
NoNewPrivileges=on
PrivateNetwork=on
PrivateTmp=on
ProtectSystem=full
ProtectHome=on

[Install]
WantedBy=sysinit.target
/tmp % ls -l | grep "rngd.service" | wc -l
15
```

possibly related systemd debug log when reloading:
```
systemd[1]: unit_file_build_name_map: normal unit file: /usr/lib/systemd/system/rngd.service
systemd[1]: rngd.service: Changed dead -> running
```

**Comment by poettering** [comment]
Please provide full debug logs for a "systemctl daemon-reload".

The code that allocates a new temporary directory is run between the fork() and the execve() in the child process for the service processes PID 1 runs. This means, if you see a new temporary directory being created then this suggests that something is restarting rngd during daemon reload.

**Comment by hexchain** [comment]
Full debug log is attached. There is also no sign indicating that rngd is restarted, `systemctl status rngd` says it is started 5 days ago.

[debug.log](https://github.com/systemd/systemd/files/5364154/debug.log)

**Comment by poettering** [comment]
can you "strace -y -s 500 -o log /tmp/strace-reload.log -f -p 1" what PID 1 does during the reload? Then attach that here.. (Doesn't have to be the full thing, it's OK to cut away unrelated stuff, where nothing relevant to rngd (and the lines around it) is done)

**Comment by poettering** [comment]
is this reproducible on other services using PrivateTmp=?

**Comment by hexchain** [comment]
> can you "strace -y -s 500 -o log /tmp/strace-reload.log -f -p 1" what PID 1 does during the reload? Then attach that here.. (Doesn't have to be the full thing, it's OK to cut away unrelated stuff, where nothing relevant to rngd (and the lines around it) is done)

[strace-reload.log](https://github.com/systemd/systemd/files/5364281/strace-reload.log)


> is this reproducible on other services using PrivateTmp=?

I don't think so:
```
% ls -lhtrd /tmp/systemd-private*
drwx------ 3 root root 60 Oct  6 23:29 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-414rgj/
drwx------ 3 root root 60 Oct  6 23:29 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-systemd-timesyncd.service-Pe5uri/
drwx------ 3 root root 60 Oct  6 23:29 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-dbus-broker.service-labSmg/
drwx------ 3 root root 60 Oct  6 23:29 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-systemd-logind.service-X3riJg/
drwx------ 3 root root 60 Oct  6 23:29 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-upower.service-vrFsQf/
drwx------ 3 root root 60 Oct  6 23:30 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-colord.service-Q4h8th/
drwx------ 3 root root 60 Oct  6 23:41 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-RQncyh/
drwx------ 3 root root 60 Oct  7 11:22 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-d4AKxg/
drwx------ 3 root root 60 Oct  7 14:59 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-EyCyBg/
drwx------ 3 root root 60 Oct  7 21:58 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-UA3VJg/
drwx------ 3 root root 60 Oct  8 22:32 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-Mvmwbj/
drwx------ 3 root root 60 Oct  9 10:44 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-wtXSWi/
drwx------ 3 root root 60 Oct 10 11:45 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-E9Y0Pi/
drwx------ 3 root root 60 Oct 10 16:01 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-TPl0Ig/
drwx------ 3 root root 60 Oct 11 15:51 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-p4MmQg/
drwx------ 3 root root 60 Oct 11 15:52 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-wU02ef/
drwx------ 3 root root 60 Oct 11 15:53 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-rDxkaj/
drwx------ 3 root root 60 Oct 11 15:53 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-rSFAUg/
drwx------ 3 root root 60 Oct 11 16:33 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-YPwYuj/
drwx------ 3 root root 60 Oct 11 16:34 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-t7Xich/
drwx------ 3 root root 60 Oct 12 11:28 /tmp/systemd-private-e43e5665639e402a8e773821a61e1b77-rngd.service-NGnOhh/
```

**Comment by hexchain** [comment]
This unit file seems to be a reproducer:

```
[Unit]
Description=Test

[Service]
ExecStart=/usr/bin/sleep infinity
PrivateNetwork=on
PrivateTmp=on
```

**Comment by poettering** [comment]
I can reproduce this in git master

**Comment by poettering** [comment]
Fix in #17351
